### PR TITLE
Models docstrings

### PR DESCRIPTION
**Pull request by A-Jacobson**
#401 
- not sure how far i should go with the `loss.py`, `initializers.py` and model architecture files. The first two probably need a good refactor, loss -> metrics and initializers -> (not sure, maybe deletion). The model architectures are copied from all over the place. Am adding docstrings but am reluctant to refactor. Our `ComposerModel` versions should be the public facing interface for these. 

- unsure about module level docstring, should i do something like this? https://docs.mosaicml.com/en/v0.3.1/models.html. or link to the incoming model cards that @ajaysaini725  is writing to cover the model descriptions.

### CODE REVIEW

**Comment by ravi-mosaicml** [comment]
> * not sure how far i should go with the `loss.py`, `initializers.py` and model architecture files. The first two probably need a good refactor, loss -> metrics and initializers -> (not sure, maybe deletion). The model architectures are copied from all over the place. Am adding docstrings but am reluctant to refactor. Our `ComposerModel` versions should be the public facing interface for these.

If a refactor is desperately needed, then let's do it. For PR review, though, could this refactor be done separately?


> * unsure about module level docstring, should i do something like this? https://docs.mosaicml.com/en/v0.3.1/models.html. or link to the incoming model cards that @ajaysaini725  is writing to cover the model descriptions.

A stub that would link to the model cards works.

**Comment by hanlint** [comment]
Thanks for putting this together! A few comments:

`ComposerModel`:
- [ ] Can you add more information to the `ComposerModel` base class docstrings? e.g. especially the coupling of forward/loss and validate/metrics.
- [ ] `forward`: its worth mentiong that the input is from the dataloader. we want to tell folks as much about our trainer loop logic as possible. same thing for loss -- say that outputs come from forward, and batch from the dataloader
- [ ] `metrics`: similarly explain that the metrics shuld be torchmetrics. The "warning" should hide the torchmetrics link (e.g. See torchmetrics[link] for more details). Mention that one can return multiple metrics that would all be evaluated on the eval_dataloader.

- [] `ComposerTransformer` docstring should also be expanded with more information about how it works and example usages
- [] same for TIMM, BERT, ResNet, GPTmodel, etc, we need some code blocks that show its usage, and more detail on how it works.
- [] TIMM has some formatting issues with `testcode` and also the `model_name` and links in the parameters. See: https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.timm.model.html#composer.models.timm.model.Timm

**Comment by hanlint** [comment]
This is ready for a re-review: cc: @hanlint @ravi-mosaicml

**Comment by ravi-mosaicml** [comment]
I took a quick look at https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.html and noticed that many model-level docstrings and `__all__` are not defined. In addition, many references are not auto-linking (set `nitpicky = True` in `conf.py` to get the full list.

Let me know when I should take another look.

**Review by ravi-mosaicml — CHANGES_REQUESTED** [review]
Requesting changes for module-level docstrings, `__all__`, and another pass on fixing broken references.

**Comment by A-Jacobson** [comment]
> I took a quick look at https://mosaicml-composer--469.com.readthedocs.build/en/469/api_reference/composer.models.html and noticed that many model-level docstrings and `__all__` are not defined. In addition, many references are not auto-linking (set `nitpicky = True` in `conf.py` to get the full list.
> 
> Let me know when I should take another look.

There are many objects in here I don't want to expose to a user such as Hparams and model objects copied from other repos that our `composermodels` wrap. I didn't write docstrings for these. I don't think I can exclude them with `__all__` either because we export them for our model registery in `trainer_hparams`. Do you have a suggestion for what to do about these?

**Comment by ravi-mosaicml** [comment]
> There are many objects in here I don't want to expose to a user such as Hparams and model objects copied from other repos that our `composermodels` wrap. I didn't write docstrings for these. I don't think I can exclude them with `__all__` either because we export them for our model registery in `trainer_hparams`. Do you have a suggestion for what to do about these?

Hparams should be exposed; #558 helped format how these are documented. Every field in an hparam class should be listed under the Params section of the docstring, but for each field, it is fine to have a one-liner that points to the underlying class. See https://docs.mosaicml.com/en/latest/api_reference/composer.callbacks.callback_hparams.html#module-composer.callbacks.callback_hparams as an example.

For model objects from other repos, are they imports, or did we copy and paste the code? Imports need not be listed in `__all__`, since one should import from the source package (not composer). If we copied and pasted the code, then we are responsible for the code now, and as such, it should have docstrings and be brought up to our standards.

**Comment by A-Jacobson** [comment]
> Hparams should be exposed; #558 helped format how these are documented. Every field in an hparam class should be listed under the Params section of the docstring, but for each field, it is fine to have a one-liner that points to the underlying class. See https://docs.mosaicml.com/en/latest/api_reference/composer.callbacks.callback_hparams.html#module-composer.callbacks.callback_hparams as an example.
> 
- Exposed and documented Hparams.

> For model objects from other repos, are they imports, or did we copy and paste the code? Imports need not be listed in `__all__`, since one should import from the source package (not composer). If we copied and pasted the code, then we are responsible for the code now, and as such, it should have docstrings and be brought up to our standards.

- Much of the code is copy and pasted and I suspect a some doesn't need to be here. I gave the main models docstrings and set some helper files to private. I'm going to have to do a larger refactor PR to bring all the copied code up to our standards.. which I do have planned. Hopefully this will do for a start.

- defined __all__ in every module + public file.

- fixed many broken links, there are still some that are broken when i run `nitpicky = True` locally. The broken ones are all external modules such as `torchmetrics.collections.MetricCollection`, builtin typing objections, torch, torchvision, and transformers objects. Are these supposed to work? I see these linked in our sphinx config. @ravi-mosaicml

**Comment by ravi-mosaicml** [comment]
> * fixed many broken links, there are still some that are broken when i run `nitpicky = True` locally. The broken ones are all external modules such as `torchmetrics.collections.MetricCollection`, builtin typing objections, torch, torchvision, and transformers objects. Are these supposed to work? I see these linked in our sphinx config. @ravi-mosaicml

Can you look into the internsphinx inventory and see what the proper path for these elements are? See https://www.sphinx-doc.org/en/master/usage/extensions/intersphinx.html#showing-all-links-of-an-intersphinx-mapping-file. OK to defer this one to later -- if so please create a seperate issue for it.

> * The `forward` method of `ComposerModel` isn't showing in the docs and I'm not sure why. Perhaps it's because we subclass `torch.nn.Module`?

Forward is excluded in the sphinx config. Would adding a `:meta public:` force it to be documented?

File: composer/models/__init__.py [hunk 1]
```diff
@@ -1,5 +1,13 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+"""The models module contains the :class:`~composer.models.base.ComposerModel` base class along with reference
+implementations of many common models. Additionally, it includes task specific convenience
+:class:`~composer.models.base.ComposerModel` s that wrap existing pytorch models with standard forward passes
+and logging to enable quick interaction with the :class:`~composer.trainer.trainer.Trainer`.
+
+See the ComposerModel docs for more details.
```
Comment (hanlint) [inline composer/models/__init__.py:8]: Should like to the docs with `:doc:` here.

File: composer/models/base.py [hunk 2]
```diff
@@ -17,38 +18,125 @@
 
 
 class ComposerModel(torch.nn.Module, abc.ABC):
-    """The minimal interface needed to use a model with :class:`composer.trainer.Trainer`."""
+    """The interface needed to use a pytorch model with :class:`~composer.trainer.trainer.Trainer`.
```
Comment (hanlint) [inline composer/models/base.py:21]:
```suggestion
    """The interface needed to use a pytorch model with :class:`.Trainer`.
```

File: composer/models/base.py [hunk 3]
```diff
@@ -17,38 +18,125 @@
...
+
+    To create a :class:`~composer.trainer.trainer.Trainer` compatible model, subclass :class:`~composer.models.base.ComposerModel` and
```
Comment (hanlint) [inline composer/models/base.py:23]:
```suggestion
    To create a :class:`.Trainer` compatible model, subclass :class:`~composer.models.base.ComposerModel` and
```

File: composer/models/base.py [hunk 4]
```diff
@@ -17,38 +18,125 @@
...
+
+            def forward(self, batch): # batch is the output of the dataloader
+                # specify how batches are passed through the model
+                inputs, _ = batch
+                return self.model(inputs)
+
+            def loss(self, outputs, batch):
+                # pass batches and `forward` outputs to the loss
+                _, targets = batch
+                return F.cross_entropy(outputs, targets)
```
Comment (hanlint) [inline composer/models/base.py:50]: Should additionally add the same as above -> For more detail, see the composer model docs (with a :doc: link)

File: composer/models/base.py [hunk 5]
```diff
@@ -17,38 +18,125 @@
...
+    """
 
     @abc.abstractmethod
-    def loss(self, outputs: Any, batch: Batch, *args, **kwargs) -> Tensors:
-        """Compute the loss of the model.
+    def forward(self, batch: Batch) -> Tensors:
+        """Compute model output given a batch from the dataloader.
+
+        .. warning:: this is method is different from vanilla pytorch `model.forward(x)` or `model(x)` as it takes a
+                     batch of data that may have to be unpacked.
```
Comment (hanlint) [inline composer/models/base.py:58]:
```suggestion
                     batch of data that has to be unpacked.
```

File: composer/models/base.py [hunk 6]
```diff
@@ -17,38 +18,125 @@
...
-            Tensors:
-                The result that is passed to :meth:`loss` as a ``Tensors``
-                object.
+            Tensors: The loss as a :class:`torch.Tensor`.
         """
         pass
 
     def metrics(self, train: bool = False) -> Metrics:
-        """Get metrics for evaluating the model.
+        """Get metrics for evaluating the model. Metrics should be torchmetrics compatible for accurate distributed
```
Comment (hanlint) [inline composer/models/base.py:124]:
```suggestion
        """Get metrics for evaluating the model. Metrics should be :class:`torchmetrics.Metrics` compatible for accurate distributed
```

File: composer/models/base.py [hunk 7]
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
...
         to all metrics returned by :meth:`metrics`.
 
+        Example:
+
+        .. code-block:: python
+
+            def validate(self, batch): # batch is the output of the dataloader
+                inputs, targets = batch
+                outputs = self.model(inputs)
+                return outputs, targets # return a tuple of (``outputs``, ``targets``)
```
Comment (hanlint) [inline composer/models/base.py:170]:
```suggestion
                return outputs, targets # return a tuple of (outputs, targets)
```

code block, not RST

File: composer/models/base.py [hunk 8]
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
...
+
+
+        How validate outputs are passed to :meth:`metrics` (pseudocode):
```
Comment (hanlint) [inline composer/models/base.py:173]:
```suggestion
        This pseudocode illustrates how validate outputs are passed to :meth:`metrics`:
```

File: composer/models/base.py [hunk 9]
```diff
@@ -58,39 +146,72 @@ def metrics(self, train: bool = False) -> Metrics:
...
-                Most often, this will be a tuple of the form (predictions, targets).
+            Tuple[Any, Any]: A Tuple of (``outputs``, ``targets``) that is passed directly to the
+                :meth:`~torchmetrics.Metric.update` methods of the metrics returned by :meth:`metrics`.
         """
         raise NotImplementedError('Implement validate in your ComposerModel to run validation.')
 
 
 class ComposerClassifier(ComposerModel):
-    """Implements the base logic that all classifiers can build on top of.
+    """A convenience class that creates a :class:`~composer.models.base.ComposerModel` for classification tasks from a
```
Comment (hanlint) [inline composer/models/base.py:196]: here and throughout should use the global scope `.ComposerModel` and `.ComposerClassifier` to make the docstrings more less verbose.

File: composer/models/bert/model.py [hunk 10]
```diff
@@ -96,13 +108,12 @@ def validate(self, batch: BatchDict) -> Tuple[Tensors, Tensors]:
 
         Args:
             batch (BatchDict): a dictionary of Dict[str, Tensor] of inputs
-                that the model expects, as found in ComposerTransformer.get_model_inputs().
+                that the model expects, as found in :meth:`composer.models.transformer_shared.ComposerTransformer.get_model_inputs`.
```
Comment (hanlint) [inline composer/models/bert/model.py:111]:
```suggestion
                that the model expects, as found in :meth:`.ComposerTransformer.get_model_inputs`.
```

File: composer/models/classify_mnist/model.py [hunk 11]
```diff
@@ -47,16 +49,28 @@ def forward(self, x):
 class MNIST_Classifier(ComposerClassifier):
     """A simple convolutional neural network extending :class:`ComposerClassifier`.
```
Comment (hanlint) [inline composer/models/classify_mnist/model.py:50]:
```suggestion
    """A simple convolutional neural network extending :class:`.ComposerClassifier`.
```

File: composer/models/classify_mnist/model.py [hunk 12]
```diff
@@ -47,16 +49,28 @@ def forward(self, x):
...
 
-    :class:`composer.models.MNIST_Classifier` is a simple example
+    :class:`~composer.models.classify_mnist.model.MNIST_Classifier` is a simple example
```
Comment (hanlint) [inline composer/models/classify_mnist/model.py:52]:
```suggestion
    A simple example
```
(self-referencing)

File: composer/models/deeplabv3/deeplabv3.py [hunk 13]
```diff
@@ -64,15 +67,24 @@ def deeplabv3_builder(num_classes: int,
 
 
 class ComposerDeepLabV3(ComposerModel):
-    """DeepLabV3 model extending the :class:`ComposerClassifier`.
+    """DeepLabV3 model extending :class:`~composer.models.base.ComposerClassifier`. Logs Mean Intersection over Union
```
Comment (hanlint) [inline composer/models/deeplabv3/deeplabv3.py:70]:
```suggestion
    """DeepLabV3 model extending :class:`.ComposerClassifier`. Logs Mean Intersection over Union
```

File: composer/models/deeplabv3/deeplabv3_hparams.py [hunk 14]
```diff
@@ -6,10 +6,13 @@
 
 from composer.models.model_hparams import ModelHparams
 
+__all__ = ["DeepLabV3Hparams"]
+
 
 @dataclass
 class DeepLabV3Hparams(ModelHparams):
-    """This class specifies arguments for a DeepLabV3 model and can instantiate a DeepLabV3 model.
+    """yahp interface for :class:`~composer.models.deeplabv3.deeplabv3.ComposerDeepLabV3`. This class specifies
```
Comment (hanlint) [inline composer/models/deeplabv3/deeplabv3_hparams.py:14]:
```suggestion
    """yahp interface for :class:`~.ComposerDeepLabV3`. This class specifies
```

**Review by hanlint — COMMENTED** [review]
This PR is a lot, thanks for doing this. Made several comments. Some major comments are:

* for brevity in the code, please use the global scope e.g. `.ComposerModel` instead of the full path `~composer....`. 
* defaults should be formatted as Default: True, not Default = True.

**Comment by A-Jacobson** [comment]
> This PR is a lot, thanks for doing this. Made several comments. Some major comments are:
> 
> * for brevity in the code, please use the global scope e.g. `.ComposerModel` instead of the full path `~composer....`.
> * defaults should be formatted as Default: True, not Default = True.

Thanks for going through it. Comments should be addressed.

**Review by hanlint — APPROVED** [review]
Agreed that the `loss` and `initializers` will need a better organization moving forward (already captured in #536 ), but that should not be blocking for this PR.

However, one major request -- for all docstrings, arguments that have defaults should be listed as optional. eg. in `Timm`, should be:

```
pretrained (bool, optional): Imagenet pretrained. Default: False.
```
Can you do a pass through all the methods in this module and add optional wherever appropriate?

File: composer/models/unet/__init__.py [hunk 15]
```diff
@@ -1,8 +1,15 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
 
+"""The Unet architecture used in image segmentation. The example we are using is for medical brain tumor data.
```
Comment (growlix) [inline composer/models/unet/__init__.py:3]: The 2nd sentence could be a bit more specific. The example dataset?

File: composer/models/unet/model.py [hunk 16]
```diff
@@ -3,10 +3,25 @@
...
 class UNet(nn.Module):
+    """Unet Architecture adapted from
+    https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/.
+
+    Args:
+        in_channels (int): number of input channels.
+        n_class (int): number of output layers.
+        kernels (list): conv layer kernel sizes.
+        strides (list): conv layer strides.
+        normalization_layer (str): normalization layer type, one of ("batch", "instance").
```
Comment (growlix) [inline composer/models/unet/model.py:20]:
```suggestion
        normalization_layer (str): normalization layer type, one of (``"batch"``, ``"instance"``).
```

File: composer/models/unet/model.py [hunk 17]
```diff
@@ -3,10 +3,25 @@
...
+        negative_slope (float): leaky relu negative slope.
+        residual (bool): use residual connections.
+        dimension (int): filter dimensions.
```
Comment (growlix) [inline composer/models/unet/model.py:23]: Capitalize first letter of parameter description

File: composer/models/unet/model.py [hunk 18]
```diff
@@ -3,10 +3,25 @@
 ## Code adapted from https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/
```
Comment (growlix) [inline composer/models/unet/model.py:3]: File-level docstring

File: composer/models/unet/unet.py [hunk 19]
```diff
@@ -1,6 +1,5 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
```
Comment (growlix) [inline composer/models/unet/unet.py:1]: File-level docstring

File: composer/models/unet/unet.py [hunk 20]
```diff
@@ -15,16 +14,20 @@
...
+    """A U-Net model extending :class:`.ComposerModel`.
 
-    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the U-Net architecture.
-    """
+    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the
+    U-Net architecture.
 
-    n_classes: Optional[int] = None
+    Args:
+        num_classes (int): The number of classes. Needed for classification tasks. Default: 3.
```
Comment (growlix) [inline composer/models/unet/unet.py:27]:
```suggestion
        num_classes (int): The number of classes. Needed for classification tasks. Default: ``3``.
```

File: composer/models/unet/unet.py [hunk 21]
```diff
@@ -15,16 +14,20 @@
...
+__all__ = ["UNet"]
+
 
 class UNet(ComposerModel):
-    """A U-Net model extending :class:`ComposerClassifier`.
+    """A U-Net model extending :class:`.ComposerModel`.
 
-    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the U-Net architecture.
-    """
+    See this `paper <https://arxiv.org/abs/1505.04597>`_ for details on the
```
Comment (growlix) [inline composer/models/unet/unet.py:23]: Reference format

File: composer/models/unet/model.py [hunk 22]
```diff
@@ -3,10 +3,25 @@
...
 
-from .layers import ConvBlock, OutputBlock, ResidBlock, UpsampleBlock
+from composer.models.unet._layers import ConvBlock, OutputBlock, ResidBlock, UpsampleBlock
+
+__all__ = ["UNet"]
 
 
 class UNet(nn.Module):
+    """Unet Architecture adapted from
+    https://github.com/NVIDIA/DeepLearningExamples/blob/master/PyTorch/Segmentation/nnUNet/.
```
Comment (growlix) [inline composer/models/unet/model.py:13]: embed link

File: composer/models/unet/unet_hparams.py [hunk 23]
```diff
@@ -1,13 +1,20 @@
 # Copyright 2021 MosaicML. All Rights Reserved.
```
Comment (growlix) [inline composer/models/unet/unet_hparams.py:1]: File-level docstring

File: composer/models/unet/unet_hparams.py [hunk 24]
```diff
@@ -1,13 +1,20 @@
...
 
+__all__ = ["UnetHparams"]
+
 
 @dataclass
 class UnetHparams(ModelHparams):
+    """yahp Hparams interface for Unet segmentation models.
+
+    Args:
+        num_classes (int, optional): The number of classes. Needed for classification tasks. Default: 3.
```
Comment (growlix) [inline composer/models/unet/unet_hparams.py:15]:
```suggestion
        num_classes (int, optional): The number of classes. Needed for classification tasks. Default: ``3``.
```

**Review by growlix — COMMENTED** [review]
99% of my comments are one of the following four comments:

1. Example and default arguments should always be formatted like ``"this"``.
2. Consistent reference formatting
3. No nude hyperlinks—embed hyperlinks in descriptive text
4. File-level docstrings

I'm sorry that we have to inflict this violence on each other. But impeccable documentation requires blood sacrifice.

File: composer/models/base.py [hunk 25]
```diff
@@ -1,54 +1,150 @@
...
-        """Compute the loss of the model.
+    def forward(self, batch: Batch) -> Tensors:
+        """Compute model output given a batch from the dataloader.
+
+        .. warning:: This method is different from vanilla PyTorch ``model.forward(x)`` or ``model(x)`` as it takes a
+                     batch of data that has to be unpacked.
+
+        Example:
+
+        .. code-block:: python
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/base.py:64]: Going to punt the doctests in the base class to another pr as they require fixtures.

File: composer/models/base.py [hunk 26]
```diff
@@ -1,54 +1,150 @@
...
-                The result that is passed to :meth:`loss` as a ``Tensors``
-                object.
+            Tensors: The loss as a :class:`torch.Tensor`.
         """
         pass
 
     def metrics(self, train: bool = False) -> Metrics:
-        """Get metrics for evaluating the model.
+        """Get metrics for evaluating the model. Metrics should be instances of :class:`torchmetrics.Metric` defined in
+        :meth:`__init__`. This format enables accurate distributed logging. Metrics consume the outputs of
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/base.py:129]: I'm referring to the sub class init here. Not sure how to link to that as it doesn't exist yet :(. Base class doesn't have an init.

File: composer/models/loss.py [hunk 27]
```diff
@@ -153,7 +159,39 @@ def soft_cross_entropy(input: Tensor,
...
                        reduction: str = 'mean'):
-    """Drop-in replacement for ``torch.CrossEntropy`` that can handle dense labels.
+    r"""Drop-in replacement for :class:`~torch.nn.CrossEntropyLoss` that can handle class indices or one-hot labels.
+
+    Args:
+        input (torch.Tensor) : :math:`(N, C)` where `C = number of classes` or :math:`(N, C, H, W)`
+            in case of 2D Loss, or :math:`(N, C, d_1, d_2, ..., d_K)` where :math:`K \geq 1`
+            in the case of K-dimensional loss. `input` is expected to contain unnormalized scores
+            (often referred to as logits).
+        target (torch.Tensor) : If containing class indices, shape :math:`(N)` where each value is
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/loss.py:169]: It should! This is the pytorch docstring with some minor modifications

File: composer/models/base.py [hunk 28]
```diff
@@ -1,54 +1,150 @@
...
+
+    To create a :class:`.Trainer` compatible model, subclass :class:`.ComposerModel` and
+    implement :meth:`forward` and :meth:`loss`. For full functionality (logging and validation), implement :meth:`metrics`
+    and :meth:`validate`.
+
+    See :doc:`Composer Model </composer_model>` for more details.
+
+    Minimal Example:
+
+    .. code-block:: python
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/base.py:31]: gonna punt this to another PR, ravi and i discussed these.

File: composer/models/base.py [hunk 29]
```diff
@@ -1,54 +1,150 @@
...
+            for batch in train_dataloader:
+                optimizer.zero_grad()
+                outputs = model.forward(batch)
+                loss = model.loss(outputs, batch)
+                loss.backward()
 
         Args:
-            batch (Batch): The input batch for the forward pass.
+            outputs (Any): The output of the forward pass.
+            batch (~composer.core.types.Batch): The output batch from dataloader.
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/base.py]: have tried a few things to fix this. was unable to. Also these are going to be removed soon anyway so I didn't want to burn too much time on it.

File: composer/models/deeplabv3/deeplabv3.py [hunk 30]
```diff
@@ -22,10 +24,11 @@ def deeplabv3_builder(num_classes: int,
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/deeplabv3/deeplabv3.py:24]: beyond    

 """Helper function to build a torchvision DeepLabV3 model with a 3x3 convolution layer and dropout removed."""
 
 ?

File: composer/models/deeplabv3/deeplabv3.py [hunk 31]
```diff
@@ -64,15 +67,25 @@ def deeplabv3_builder(num_classes: int,
 
 
 class ComposerDeepLabV3(ComposerModel):
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/deeplabv3/deeplabv3.py:69]: only that I didn't want to add a refactor to this already huge PR. see https://github.com/mosaicml/composer/issues/536

File: composer/models/base.py [hunk 32]
```diff
@@ -58,39 +154,72 @@ def metrics(self, train: bool = False) -> Metrics:
...
-            batch (Batch): The data to perform validation with.
-                Specified as a tuple of tensors (input, target).
+            batch (~composer.core.types.Batch): The output batch from dataloader
 
         Returns:
-            Tuple[Any, Any]: Tuple that is passed directly to the
-                `update()` methods of the metrics returned by :meth:`metrics`.
-                Most often, this will be a tuple of the form (predictions, targets).
+            Tuple[Any, Any]: A Tuple of (``outputs``, ``targets``) that is passed directly to the
+                :meth:`~torchmetrics.Metric.update` methods of the metrics returned by :meth:`metrics`.
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/base.py]: yeah a lot of torchmetrics links are broken. The package authors don't have working links of it either. https://torchmetrics.readthedocs.io/en/latest/references/modules.html?highlight=Metrics . not sure what to do about this.

File: composer/models/bert/model.py [hunk 33]
```diff
@@ -19,16 +19,27 @@
...
+        hf_model = transformers.BertLMHeadModel(config=config)
+        tokenizer = transformers.BertTokenizer.from_pretrained("bert-base-uncased")
+        model = BERTModel(module=hf_model, config=config, tokenizer=tokenizer)
 
     Args:
-        module (transformers.BertModel): The model to wrap with this module.
-        config (transformers.BertConfig): The config for the model.
-        tokenizer (transformers.BertTokenizer): The tokenizer used for this model,
-            necessary to assert required model inputs.
+        module (transformers.BertModel): An instance of BertModel that
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/bert/model.py]: same as torchmetrics, I believe transformers doesn't have a compatible RTD table i'm not sure how to get them to work.

File: composer/models/nlp_metrics.py [hunk 34]
```diff
@@ -78,7 +81,7 @@ def update(self, output: Union[Mapping, Tensor], target: Tensor) -> None:
         Args:
             output (Mapping): The output from the model, which must contain
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/nlp_metrics.py:82]: Have tried a few path variations with no luck. any ideas on this?

File: composer/models/nlp_metrics.py [hunk 35]
```diff
@@ -54,9 +57,9 @@ class CrossEntropyLoss(Metric):
 
     Args:
         vocab_size (int): the size of the tokenizer vocabulary.
-        dist_sync_on_step (bool): Synchronize metric state across processes at
+        dist_sync_on_step (bool, optional): Synchronize metric state across processes at
             each forward() before returning the value at the step.
-        ignore_index (int): The class index to ignore. Defaults to -100.
+        ignore_index (int, optional): The class index to ignore. Defaults to -100.
 
     State:
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/nlp_metrics.py:64]: saw the slack discussion on this but it doesn't look like there's a great solution. either set as attributes or ignore it for now?

**Comment by A-Jacobson** [comment]
Done:

- Added all file level docstrings.
- embedding links.
- added all the `` marks...everywhere.
- vit docstrings, sorta minimal.
- bare minimum ssd docstrings.
- refactored effecientnet to hide non user-facing blocks.

Not Done:
- some links are still broken: transformers, base types, some torchmetrics. These broken links don't have readthedocs paths available for us to you. I've linked to the relevant documentation or github pages for some. For other's i've left alone either because we're working to remove them (Batch types), or they're in an awkward place to add a normal link (Return or Args). Open to suggestions here but I do think it's ok to send this off without them.
- codeblocks -> doctests in abstract classes, punting to a later PR as they require fixtures + some are psuedocode.
- `State` formatting on metrics, adding them as attrs will cause more trouble than living with the less pretty formatting for now.
- SSD is a lot.. I know it's not perfect but I'd like to get this PR moving before even more models appear.

@growlix I choose you!

**Comment by growlix** [comment]
Looking greaet, Austin. Hope your eyes/fingers aren't too bloody.

> embedded paper links, they are cited as [Name](Link) not [Name](link)(Author et al, Date). Should i do (Author et al, Date)? Is there a relatively painless way to get this information?

I'd go [Name](link)(Author et al, Date) if possible; [Author et al, Date](link) is also acceptable. I think author attribution is the most important part. Unfortunately I don't know if there's an easy way to do this other than clicking through to the actual page.

> some links are still broken: transformers, base types, some torchmetrics. These broken links don't have readthedocs paths available for us. I've linked to the relevant documentation or github pages for some. I've left others alone either because we're working to remove them (Batch types), or they're in an awkward place to add a normal link (Return or Args). Open to suggestions here but I do think it's ok to send this off without them.

Agree. I think you can leave as-is.

> codeblocks -> doctests in abstract classes, punting to a later PR as they require fixtures + some are psuedocode.

Sounds good, just remember to file an issue.

> SSD is a lot.. I know it's not perfect but I'd like to get this PR moving before even more models appear.

I'll double check for anything egregious, but otherwise sounds good, file an issue.

> State formatting on metrics, adding them as attrs will cause more trouble than living with the less pretty formatting for now.

As per [Bandish's suggestion](https://mosaicml.slack.com/archives/C02F05H39SL/p1646773237565229?thread_ts=1646768731.433729&cid=C02F05H39SL), I think you can just move the state info to a new paragraph in the summary and preface it with something like

```
Adds the following attributes to :class:`composer.core.state`:
```
(Don't quote my syntax on that)

Unfortunately I think the current rendering is a bit too sloppy.

> general refactoring (standardizing module naming and layout) is being punted to here, https://github.com/mosaicml/composer/issues/536. This PR is already too crazy, lets end the pain.

+ 1

**Comment by A-Jacobson** [comment]
- author attribution in.
- state information moved above params on `nlp_metrics` not sure this is prettier but it's done.
- issue filed for codeblocks -> testcode https://github.com/mosaicml/composer/issues/703
- issue filed for ssd cleanup https://github.com/mosaicml/composer/issues/704

**Review by growlix — APPROVED** [review]
SSD files need `__all__`, but otherwise looks p good

File: composer/models/ssd/ssd.py [hunk 36]
```diff
@@ -1,3 +1,5 @@
+"""Single Shot Object detection Model with pretrained ResNet34 backbone extending :class:`.ComposerModel`."""
```
[the remark this answers is not in the document]
Comment (A-Jacobson) [reply composer/models/ssd/ssd.py:1]: im cool with it.

**Pull request merged**